src/FffGcodeWriter.cpp

```diff
@@ -1211,6 +1211,356 @@ void FffGcodeWriter::processSpiralizedWall(const SliceDataStorage& storage, Laye
+    const bool outer_inset_first = mesh->getSettingBoolean("outer_inset_first")
+                                    || (layer_nr == 0 && mesh->getSettingAsPlatformAdhesion("adhesion_type") == EPlatformAdhesion::BRIM);
+    const unsigned num_insets = part.insets.size();
+    constexpr bool spiralize = false;
+    constexpr float flow = 1.0;
+    // create a vector of vectors containing all the inset polys
+    std::vector<std::vector<ConstPolygonRef>> inset_polys;
+    for (unsigned inset_idx = 0; inset_idx < num_insets; ++inset_idx)
+    {
+        inset_polys.emplace_back();
+        for (unsigned poly_idx = 0; poly_idx < part.insets[inset_idx].size(); ++poly_idx)
+        {
+            inset_polys[inset_idx].push_back(part.insets[inset_idx][poly_idx]);
+        }
+    }
+    // work out the order we wish to visit all the holes (doesn't include the outer wall of the part)
+    PathOrderOptimizer orderOptimizer(gcode_layer.getLastPosition(), z_seam_pos, z_seam_type);
+    for (unsigned int poly_idx = 1; poly_idx < inset_polys[0].size(); poly_idx++)
+    {
+        orderOptimizer.addPolygon(inset_polys[0][poly_idx]);
+    }
+    orderOptimizer.optimize();
+    // this will consume all of the insets that surround holes but not the insets next to the outermost wall of the model
+    for (unsigned outer_poly_order_idx = 0; outer_poly_order_idx < orderOptimizer.polyOrder.size(); ++outer_poly_order_idx)
+    {
+        Polygons hole_outer_wall; // the outermost wall of a hole
+        hole_outer_wall.add(inset_polys[0][orderOptimizer.polyOrder[outer_poly_order_idx] + 1]); // +1 because first element (part outer wall) wasn't included
+        int adjacent_poly_idx = -1; // index of poly in inset_polys[1] that is adjacent (ideally enclosing) the hole
+        if (inset_polys.size() > 1)
+        {
+            // find the adjacent poly in the level 1 insets that encloses the hole
+            adjacent_poly_idx = findAdjacentEnclosingPoly(hole_outer_wall[0], inset_polys[1], std::max(wall_line_width_0, wall_line_width_x) * 1.5f);
+            if (adjacent_poly_idx < 0 && !outer_inset_first)
+            {
+                // we didn't find a level 1 inset that encloses this hole so now look to see if there is a level 1 inset that simply touches
+                // this hole and use that instead - however, as the level 1 inset will touch other holes and/or the outer wall we don't want
+                // to do this when printing the outer walls first
+                adjacent_poly_idx = findAdjacentPoly(hole_outer_wall[0], inset_polys[1], std::max(wall_line_width_0, wall_line_width_x) * 1.5f);
+            }
+        }
+        // now test for the special case where we are printing the outer wall first and we have two or more holes so close together that they share a level 1 inset
+        // in this situation we don't want to output the level 1 inset until after all the holes' outer walls have been printed
```

Comment:
The inverse case is also problematic.
You can have a big 8-shaped outline which has two second walls for the outside part, but you can also have a big part with two round holes close to each other, such that the second wall around those two holes forms an 8-shape.

In the first case we have a problem when we need to process the inner inset first, while in the second case we have a problem when we need to process the outer inset first.